docs/development/design/future-builder.rst

```diff
@@ -0,0 +1,230 @@
+* Translate our current magic at build time to a defined contract with the user
+* Provide a way to add a command argument without implementing it as a config file (e.g. ``fail_on_warning``)
+* Define a path forward towards supporting other tools
+* Re-write all ``readthedocs-sphinx-ext`` features to post-processsing HTML features
+* Reduce complexity maintained by Read the Docs' core team
+* Make Read the Docs responsible for Sphinx support and delegate other tools to the community
+* Eventually support upload pre-build docs
+* Allow us to add a feature with a defined contract without worry about breaking old builds
+* Introduce ``build.builder: 2`` config (does not install pre-defined packages) for these new features
+* Motivate users to migrate to ``v2`` to finally deprecate this magic by educating users
+
+
+Steps ran by the builder
+------------------------
+
+Read the Docs currently controls all the build process.
+Users are only allowed to modify very limited behavior by using a ``.readthedocs.yaml`` file.
+This drove us to implement features like ``sphinx.fail_on_warning``, ``submodules``, among others,
+at a high implementation and maintenance cost to the core team.
+Besides, this hasn't been enough for more advanced users that require more control over these commands.
+
+This document proposes to clearly define the steps the builder ran and allow users to override them
+depending on their needings:
+
+- Newbie user / simple platform usage: Read the Docs controls all the commands (current builder)
+- Intermediate user: ability to override one or more commands plus running pre/post hooks
+- Advanced user: controls *all the commands* executed by the builder
+
+The steps identified so far are:
+
+#. Checkout
+#. Expose project data via environment variables (\*)
+#. Create environment (virtualenv / conda)
+#. Install dependencies
+#. Build documentation
+#. Generate defined contract (``metadata.yaml``)
+#. Post-process HTML (\*)
+#. Upload to storage (\*)
+
+Steps marked with *(\*)* are managed by Read the Docs and can't be overwritten.
+
+
+Defined contract
+----------------
+
+Projects building on Read the Docs must provide a ``metadata.yaml`` file after running their last command.
+This file contains all the data required by Read the Docs to be able to add its integrations.
+If this file is not provided or malformed, Read the Docs will fail the build and stop the process
+communicating to the user that there was a problem with the ``metadata.yaml`` and we require them to fix the problem.
+
+.. note::
+
+   There is no restriction about how this file is generated
+   (e.g. generated with Python, Bash, statically uploaded to the repository, etc)
+   Read the Docs does not have control over it and it's only responsible for generating it when building with Sphinx.
+
+
+The following is an example of a ``metadata.yaml`` that is generated by Read the Docs when building Sphinx documentation:
+
+.. code:: yaml
+
+   # metadata.yaml
```

Comment:
Feels like all this information can be in the rtd config file, I don't see why you would want to make this dynamically generated to change at runtime.

Reply:
I put some static data here just as an example. However, I'm reserving this file to act as a place to specify dynamic metadata that's only known at build time by the doctool: something that it's known _after_ the tool has built the documentation or _while_ it's building it.

I'm proposing to create the pipeline in a generic way to be prepared for these cases. It may not be needed to generate this file dynamically for some tools/users but it must be required for other tools (*).

Besides, I want to separate it from `readthedocs.yaml` since I imagine this could eventually be useful for uploading pre-built documentation where you upload a `project-html.zip` + `metadata.yaml` and Read the Docs gets everything needed from the `metadata.yaml` without the needing of a `readthedocs.yaml` (which is currently only useful to configure the build process and we won't have one in this case).

I also envision `readthedocs.yaml` as a way to "configure the build process in the platform" (in the future we may be able o configure other platform features than just the build, tho). On the other hand, I see `metadata.yaml` as an "integration point between the build's output and Read the Docs' features/integrations (flyout, search, warning banners, hosting, etc)"

> (*) For example, the PDF filename in Sphinx is generated based on the `latex_documents` config (and `-jobname` argument in `latexmk` command) and it would be handy to modify it in one place and keep everything working without manually updating the `readthedocs.pdf_output` field in the `metadata.yaml`